util/topsql/reporter/reporter.go

```diff
@@ -654,11 +648,16 @@ func (tsr *RemoteTopSQLReporter) doReport(data *ReportData) {
 	})
 	deadline := time.Now().Add(timeout)
 
-	tsr.dataSinks.mu.Lock()
-	defer tsr.dataSinks.mu.Unlock()
-	for ds := range tsr.dataSinks.m {
+	tsr.dataSinkMu.Lock()
+	for ds := range tsr.dataSinks {
+		tsr.tmpDataSinks = append(tsr.tmpDataSinks, ds)
+	}
+	tsr.dataSinkMu.Unlock()
+
+	for _, ds := range tsr.tmpDataSinks {
 		if err := ds.TrySend(data, deadline); err != nil {
 			logutil.BgLogger().Warn("[top-sql] failed to send data to datasink", zap.Error(err))
 		}
 	}
+	tsr.tmpDataSinks = tsr.tmpDataSinks[:0]
```

Comment:
There is data race for this assignment. Your `tmpDataSinks` should be function scoped so that there is no race when accessing it at all.